src/ServiceControl/Monitoring/RavenDbMonitoringDataStore.cs

```diff
@@ -119,6 +119,27 @@ public async Task Delete(Guid endpointId)
             }
         }
 
+        public async Task BulkCreate(EndpointDetails[] endpoints)
+        {
+            using (var session = store.OpenAsyncSession())
```

Comment:
If I understood correctly the assumption here is that we will never create a lot of endpoints and therefore moving it out from bulk session is OK. 

We got some great insights from the calls with Hibernating Rhinos talking about RavenDB 5. Especially this part

> When moving to RavenDB 5 we should move away from BulkInserts. BulkInserts are faster in RavenDB 3.5 because RavenDB acquired a global lock when doing session operations. With bulk inserts, that problem is less exacerbated. The suggestion was to have a pattern of interleaved operations feeding from the transport and shuffling to RavenDB. Since naturally our batch sizes are quite small bulk inserts with RavenDB are not worth it. Instead we should for example split the "concurrency" in half and then have a few RavenDB sessions that store multiple operations and then call SaveChanges. More or less the following pattern

this would indicate we will have to move away from doing BulkInsert and actually have one document session wrapping everything that happens during ingestion. That is also a good pattern in general with SQL Server. Wouldn't it make more sense under that light to introduce a unit of work pattern that opens a storage session and saves it at the end and have all "plugins" rely on that being around via DI or some explicit mechanism like we treat the synchronized storage in core (I would prefer that)? Then we write today code that is more future proof because

a) we can move the bulk session handling into the unit of work as well
b) for the changes here just create another session alongside with it

once we move to v5 we can just delete the bulk session handling and rely on the session only. I think this pattern of starting to have individual plugins doing individual connection handling can be dangerous on the long run because it is easy to create all sorts of plugins that do "micro" transactions against the underlying storage that causes the ingestion rate to drop significantly.

Reply:
Sounds reasonable